libraries/AC_AttitudeControl/AC_AttitudeControl_Heli.cpp

```diff
@@ -539,29 +539,45 @@ void AC_AttitudeControl_Heli::set_throttle_out(float throttle_in, bool apply_ang
     _throttle_in = throttle_in;
     update_althold_lean_angle_max(throttle_in);
 
-    if (_transition_count > 0) {
-        _transition_count -= 1;
+    _motors.set_throttle_filter_cutoff(filter_cutoff);
+    if (apply_angle_boost && !((AP_MotorsHeli&)_motors).get_in_autorotation()) {
+        // Apply angle boost
+        throttle_in = get_throttle_boosted(throttle_in);
     } else {
-        _transition_count = 0;
+        // Clear angle_boost for logging purposes
+        _angle_boost = 0.0f;
     }
-    float throttle_out = 0.0f;
-    if (_transition_count > 0) {
-        if ((_ahrs.roll_sensor >= -3000 && _ahrs.roll_sensor <= 3000) || _ahrs.roll_sensor >= 15000 || _ahrs.roll_sensor <= -15000) {
-            throttle_out = (throttle_in - ((AP_MotorsHeli&)_motors).get_coll_mid()) / cosf(radians(_ahrs.roll_sensor * 0.01f)) + ((AP_MotorsHeli&)_motors).get_coll_mid();
-        } else if ((_ahrs.roll_sensor > 3000 && _ahrs.roll_sensor < 15000) || (_ahrs.roll_sensor > -15000 && _ahrs.roll_sensor < -3000)) {
-            float scale_factor = cosf(radians(_ahrs.roll_sensor * 0.01f)) / cosf(radians(30.0f));
-            throttle_out = scale_factor * (throttle_in - ((AP_MotorsHeli&)_motors).get_coll_mid())/ cosf(radians(30.0f)) + ((AP_MotorsHeli&)_motors).get_coll_mid();
-        }
-    } else if (_inverted_flight) {
-        throttle_out = 1.0f - throttle_in;
-    } else {
-        throttle_out = throttle_in;
+    _motors.set_throttle(throttle_in);
+}
+
+// returns a throttle including compensation for roll/pitch angle
+// throttle value should be 0 ~ 1
+float AC_AttitudeControl_Heli::get_throttle_boosted(float throttle_in)
+{
+    if (!_angle_boost_enabled) {
+        _angle_boost = 0;
+        return throttle_in;
     }
+    // inverted_factor is 1 for tilt angles below 60 degrees
+    // inverted_factor changes from 1 to -1 for tilt angles between 60 and 120 degrees
+
+    float cos_tilt = _ahrs.cos_pitch() * _ahrs.cos_roll();
+    float inverted_factor = constrain_float(2.0f * cos_tilt, -1.0f, 1.0f);
+    float cos_tilt_target = fabsf(cosf(_thrust_angle));
+    float boost_factor = 1.0f / constrain_float(cos_tilt_target, 0.1f, 1.0f);
+
+    // angle boost and inverted factor applied about the zero thrust collective
+    float throttle_out = (throttle_in - ((AP_MotorsHeli&)_motors).get_coll_mid()) * inverted_factor * boost_factor + ((AP_MotorsHeli&)_motors).get_coll_mid();
```

Comment:
```suggestion
    const float coll_mid = (AP_MotorsHeli&)_motors).get_coll_mid();
    float throttle_out = ((throttle_in - coll_mid)  * inverted_factor * boost_factor) + coll_mid;
```
Pulling out the scary looking motors type cast makes it easier to follow I think.